h2/src/main/org/h2/command/Parser.java

```diff
@@ -6360,15 +6360,23 @@ private void readString(int i, char[] chars, int[] types) {
     }
 
     private String readRawString(int i, char[] chars, int[] types) {
-        StringBuilder result = new StringBuilder();
+        String result = null;
+        StringBuilder builder = null;
         for (;; i++) {
             boolean next = false;
             for (;; i++) {
                 int begin = i;
                 while (chars[i] != '\'') {
                     i++;
                 }
-                result.append(sqlCommand, next ? begin - 1 : begin, i);
+                if (result == null) {
+                    result = sqlCommand.substring(begin, i);
+                } else {
+                    if (builder == null) {
+                        builder = new StringBuilder(result);
```

Comment:
might possibly be more efficient here to construct the StringBuilder with size

       result.size() + sqlCommand.size()

to avoid some re-allocations

Reply:
That's too much. `new StringBuilder(String)` allocates space for 16 additional characters, so the next addition in many cases don't create a reallocation.

Actually these lines of core aren't executed for the most of character string literals. Literals with many parts (`'something ' 'something ' 'something'`) are rarely used, many programmers and DBAs don't even know that they exists, and only few literals have `'` character in them (`' '' '`).

Reply:
I thought about `result.size() + i - begin + 1` or something like it, but if string has one `'`, it can have more, and we also can find a multi-part literal. So such optimization also doesn't look good, that's why I decided to use the default behavior of `StringBuilder`.